### ISSUE TITLE
Refine schema requirements for feature (gene) names. 

### ISSUE THREAD

**Issue by ambrosejcarr** [chanzuckerberg/corpora-data-portal#523]
Appetite: 7

Explore ways that we could update the schema to meet the following requirements: 

- Scientists MUST be able to explore and download datasets whose features are indexed by unique, human readable symbols
- We MUST be able to capture both raw and processed data from scientists
- Our approach MUST be able to handle non-gene (e.g. protein or splice variant) features
- Our approach SHOULD minimize additional work that scientists must do to meet our requirements (e.g. we should avoid deviations from standard workflows)

**Comment by mckinsel** [comment]
We also must handle some non-human genes too. We have datasets that use human + SARS-CoV-2 as the reference, for example.

**Comment by mckinsel** [comment]
Oh and I've kind of thrown in the towel on MGI, let alone all the NHPs we've been getting. MGI seems to lack the symbol history that you get in HGNC, and there's a long tail of lemur, chimp, marmoset, etc etc etc that's not probably worth tracking down.

**Comment by mckinsel** [comment]
After experimenting with applying this to the M1 datasets, this is where I believe we've landed:

First, we need to make sure that the var.index is human readable. This is an important requirement for cellxgene to be useful.
```
if the organism is human:
    if var.index is not HGNC symbols:
        if one of the columns in var is HGNC symbols:
            swap var.index with the HGNC column
        elif there is a known mapping from var.index to HGNC symbols (certain widely used ids for example):
            apply the mapping and use the result as var.index
        else:
            fail back to the submitter
else:
    if var.index is not human readable:
        if any of the columns in var is human readable:
            swap var.index with the human readable column
        else:
            fail back to the submitter
```
"human readable" is not defined there, but in practice it's not ambiguous.

In the case of HGNC symbols (which based on the above should be all human datasets), we'd also like those symbols to be as consistent as possible across datasets. So we apply "upgrade" logic to the symbols using the previous/alias logic from HGNC.

The above processes may introduce duplicate values in var.index, or they might already be there from the submitter. This definitely breaks cellxgene and has to get fixed before we can host the data. The fix is to combine the values from the duplicate indices, but that's kind of tricky, since the correct `combine(x, y)` depends on what kind of values x and y are. We had thought that the way to resolve this was to get the raw counts, merge the raw counts by letting `combine = sum`, and then "replaying" all the normalization/transformation steps to generate any other data layers. But replaying those steps is very hard as there are a lot of differences in methods, and submitters sometimes cannot completely describe the process. It's hard enough that we were considering loosening up the gene symbol requirements so we wouldn't have to discard too many datasets.

However, I think we can get away with something a little simpler. In the datasets we've looked at thus far, every layer is one of count, log1p of `f(counts)`, or sqrt of `f(counts)`. You can verify this by applying the inverse elementwise and seeing if the cell sums look normalized. When that's the case, we can get away with not knowing everything that was done to transform and just do something like `log1p(merge_by_sum(expm1(matrix)))`, or `sqrt` and `square`. This isn't hard, so we shouldn't lose many datasets with the gene requirement, and we can keep it in the schema.

The one exception I haven't looking into yet is SCTransform. I'm not sure we can get away with this approach there.

**Comment by ambrosejcarr** [comment]
Can you look into SCTransform? It is the recommended Seurat workflow, and based on that I expect it will increase in use. If it expedites, I can connect you with the author directly on cziscience slack.

**Comment by ambrosejcarr** [comment]
> Our approach MUST be able to handle non-gene (e.g. protein or splice variant) features

I'm realizing that this is a larger task than we should tackle in this ticket. I've created chanzuckerberg/single-cell#56 to track support for other data modalities. If your research discovered relevant information about how to handle other modalities, please add it as a comment to that epic or the appropriate child.

**Comment by mckinsel** [comment]
@ambrosejcarr we now have a test case for SCTransform in https://github.com/chanzuckerberg/cellxgene/pull/1852 . Given that we want to push the broader non-gene features into a separate ticket, okay with closing this?

**Comment by ambrosejcarr** [comment]
Yes, if you link the other ticket so we can follow the trail, I think it's
fine to close.

**Comment by mckinsel** [comment]
Okay sounds good. I think it is linked to the epic...but lmk if I'm misunderstanding that.
